### PR TITLE
*: support pushing down stream aggregation on mocktikv

### PR DESCRIPTION
**Pull request by zimulala**

### CODE REVIEW

File: plan/gen_physical_plans.go [hunk 1]
```diff
@@ -22,8 +22,12 @@ import (
 	"github.com/pingcap/tidb/types"
 )
 
+// SupportStreamAggPushDown will be removed after supporting stream aggregation on TiKV.
+// TODO: Remove it.
+var SupportStreamAggPushDown bool
```
Comment (hanfei1991) [inline plan/gen_physical_plans.go]: Don't use a global variable to control stream agg push down. We should check if the client supports it. See https://github.com/pingcap/tidb/blob/master/store/tikv/coprocessor.go#L45

File: plan/gen_physical_plans.go [hunk 2]
```diff
@@ -490,26 +494,35 @@ func (p *LogicalAggregation) getStreamAggs(prop *requiredProp) []PhysicalPlan {
...
 		}
-		childProp := &requiredProp{
-			cols:        keys,
-			desc:        prop.desc,
-			expectedCnt: prop.expectedCnt * p.inputCount / p.stats.count,
-		}
-		if !prop.isPrefix(childProp) {
-			continue
+		for _, tp := range wholeTaskTypes {
+			if tp == copDoubleReadTaskType {
```
Comment (hanfei1991) [inline plan/gen_physical_plans.go]: comment why double read is not allowed

File: store/tikv/mocktikv/aggregate.go [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+	e.nextGroupByRow = e.tmpGroupByRow
+	return !firstGroup, nil
+}
+
+func (e *streamAggExec) Next(goCtx goctx.Context) (retRow [][]byte, err error) {
+	if e.executed {
+		return nil, nil
+	}
+
+	if e.aggCtxs == nil {
```
Comment (hanfei1991) [inline store/tikv/mocktikv/aggregate.go]: This should be check in building phase.

File: store/tikv/mocktikv/aggregate.go [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+	}
+	if matched {
+		return false, nil
+	}
+	e.nextGroupByRow = e.tmpGroupByRow
+	return !firstGroup, nil
+}
+
+func (e *streamAggExec) Next(goCtx goctx.Context) (retRow [][]byte, err error) {
+	if e.executed {
```
Comment (hanfei1991) [inline store/tikv/mocktikv/aggregate.go]: Why do we need `executed` ?

File: store/tikv/mocktikv/aggregate.go [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+}
+
+func (e *streamAggExec) meetNewGroup(row [][]byte) (bool, error) {
+	if len(e.groupByExprs) == 0 {
+		return false, nil
+	}
+
+	e.tmpGroupByRow = e.tmpGroupByRow[:0]
+	matched, firstGroup := true, false
+	if e.nextGroupByRow == nil {
```
Comment (hanfei1991) [inline store/tikv/mocktikv/aggregate.go]: Should we use `currGroupByRow` to check ?

File: store/tikv/mocktikv/aggregate.go [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+type streamAggExec struct {
+	evalCtx           *evalContext
+	aggExprs          []aggregation.Aggregation
+	aggCtxs           []*aggregation.AggEvaluateContext
+	groupByExprs      []expression.Expression
+	relatedColOffsets []int
+	row               types.DatumRow
+	tmpGroupByRow     types.DatumRow
+	currGroupByRow    types.DatumRow
```
Comment (hanfei1991) [inline store/tikv/mocktikv/aggregate.go]: I think currGroupByRow should be previous gby row, and next Group by row should be current gby row,

File: store/tikv/mocktikv/aggregate.go (hunk 5 above)
Comment (zimulala) [reply store/tikv/mocktikv/aggregate.go]: I think both are OK.

File: store/tikv/mocktikv/aggregate.go (hunk 4 above)
Comment (zimulala) [reply store/tikv/mocktikv/aggregate.go]: Make sure the executor is finished.

**Comment by shenli** [comment]
@hanfei1991 PTAL

File: plan/gen_physical_plans.go [hunk 7]
```diff
@@ -543,26 +544,41 @@ func (p *LogicalAggregation) getStreamAggs(prop *requiredProp) []PhysicalPlan {
...
 		}
-		childProp := &requiredProp{
-			cols:        keys,
-			desc:        prop.desc,
-			expectedCnt: prop.expectedCnt * p.inputCount / p.stats.count,
-		}
-		if !prop.isPrefix(childProp) {
-			continue
+		for _, tp := range wholeTaskTypes {
+			// Table scan can't meet the stream aggregation's require prop.
```
Comment (winoros) [inline plan/gen_physical_plans.go]: `Second read in the double read can't meet ....` may be more clear

**Pull request merged**